### ISSUE TITLE
feature request: non-transactional DDL for concurrent index creation

### ISSUE THREAD

**Issue by tgriesser** [koskimas/kysely#352]
👋 First just wanted to say this project looks awesome! Pretty close to what I wish Knex was if I started it now, with the affordances of modern TypeScript instead of 10 years ago when we were stuck in ES5 and Promises weren't even really an official language construct yet.

I started to play around with it a bit and couldn't find a way to do non-transactional DDL, though there's a chance I was missing something. This is necessary when running statements like `CREATE INDEX CONCURRENTLY` which [cannot be run inside a transaction](https://www.postgresql.org/docs/current/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY) in PostgreSQL.

Knex does this with a `config` property that can be set in a transaction file:

https://knexjs.org/guide/migrations.html#transactions-in-migrations

```ts
export const config = { transaction: false }
```

Wanted to see if this was something on the radar or if you'd be open to a similar API for configuring per-file migrations, or another approach that'd offer similar support.

**Comment by koskimas** [comment]
Hi, and thank you!

I've used knex for a long time and both of my OSS projects owe a lot of their success to knex.

> with the affordances of modern TypeScript instead of 10 years ago when we were stuck in ES5 and Promises weren't even really an official language construct yet.

Yeah, objection suffers from the same problem. It was also designed before typescript and good typescript support would mean almost a full rewrite.

About the issue itself: seems like a good (and necessary) addition.

**Comment by volnei** [comment]
What about it?

**Comment by antogyn** [comment]
Hey 👋! I'm just starting out with Kysely, and this is a blocker for me. In my case, indexes sometimes need to be created concurrently in some production environments, so Kysely can't be used for migrations - at least not for creating these specific indexes.

To expand on the original post, [Knex also has a global option "disableTransactions"](https://knexjs.org/guide/migrations.html#migration-api).

**Comment by reidswan** [comment]
We also have the use case of needing indexes to be created concurrently in production. 

I'd be happy to pitch in with a PR for this if need be. Not sure what the process for deciding on an interface for this is though; but an idea is to add an `enableTransactionalDdl` option on the [PostgresDialect](https://github.com/kysely-org/kysely/blob/master/src/dialect/postgres/postgres-dialect-config.ts#L6) and [PostgresAdapter](https://github.com/kysely-org/kysely/blob/master/src/dialect/postgres/postgres-adapter.ts), and then the [supportsTransactionalDdl](https://github.com/kysely-org/kysely/blob/master/src/dialect/postgres/postgres-adapter.ts#L10) method can just return that value?

**Comment by igalklebanov** [comment]
Hey @reidswan 👋 

Thanks for offering to help! 🙏 

I'm afraid that an option on a dialect is not the way to tackle this. However, it is a reasonable temporary workaround for anyone who needs to run migrations without a transaction using a dialect that otherwise will force a transaction.

PR's welcome to add `disableTransactions` as a `Migrator` option. Wanna be assigned to this?

**Comment by reid-stitch** [comment]
Hi @igalklebanov 

That sounds good. Yes, I'd like to be assigned to this 🙏

**Comment by reidswan** [comment]
Hi @igalklebanov 

That sounds good. Yes, please assign this to me 🙏

**Comment by igalklebanov** [comment]
@reidswan if you have any questions, swing by the discord "contributing" channel or DM me.

**Comment by fjorgemota** [comment]
#1335 was merged and `disableTransactions` is available in a released version..

Shouldn't we close this? :)

(unless I'm missing something..)

**Comment by lourd** [comment]
Not quite! Being able to run migrations in a transaction but allow for individual migrations to opt out, per the original comment, is still desired and not implemented yet.

**Comment by lourd** [comment]
I decided to take a swing at this in https://github.com/kysely-org/kysely/pull/1671, please let me know what you guys think! 🙂
